### ISSUE TITLE
[CI] TokenAuthIntegTests testRefreshingMultipleTimesWithinWindowSucceeds failing

### ISSUE THREAD

**Issue by albertzaharovits** [elastic/elasticsearch#85697]
Failed only 7 times in the last 28 days, and started failing in the last week or so:
https://gradle-enterprise.elastic.co/scans/tests?search.startTimeMax=1649167149683&search.startTimeMin=1646085600000&search.timeZoneId=Europe/Bucharest&tests.container=org.elasticsearch.xpack.security.authc.TokenAuthIntegTests&tests.sortField=FAILED&tests.test=testRefreshingMultipleTimesWithinWindowSucceeds&tests.unstableOnly=true

And sometimes the assertion is:
```
java.lang.AssertionError: 	
Expected: <1>	
     but: was <3>
```

**Build scan:**
https://gradle-enterprise.elastic.co/s/oit7vnoh6abfs/tests/:x-pack:plugin:security:internalClusterTest/org.elasticsearch.xpack.security.authc.TokenAuthIntegTests/testRefreshingMultipleTimesWithinWindowSucceeds

**Reproduction line:**
`./gradlew ':x-pack:plugin:security:internalClusterTest' --tests "org.elasticsearch.xpack.security.authc.TokenAuthIntegTests.testRefreshingMultipleTimesWithinWindowSucceeds" -Dtests.seed=F7EB4530E2A3CA68 -Dbuild.snapshot=false -Dtests.jvm.argline="-Dbuild.snapshot=false" -Dtests.locale=es-PA -Dtests.timezone=America/Rankin_Inlet -Druntime.java=17`

**Applicable branches:**
master, 8.2

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.security.authc.TokenAuthIntegTests&tests.test=testRefreshingMultipleTimesWithinWindowSucceeds

**Failure excerpt:**
```
java.lang.AssertionError: 
Expected: <1>
     but: was <2>

  at __randomizedtesting.SeedInfo.seed([F7EB4530E2A3CA68:D6E525F2AF664A1C]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.junit.Assert.assertThat(Assert.java:923)
  at org.elasticsearch.xpack.security.authc.TokenAuthIntegTests.testRefreshingMultipleTimesWithinWindowSucceeds(TokenAuthIntegTests.java:706)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:568)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:824)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:475)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:831)
  at java.lang.Thread.run(Thread.java:833)

```

**Comment by albertzaharovits** [comment]
Not muting for now.

**Comment by dimitris-athanasiou** [comment]
Another failure in https://gradle-enterprise.elastic.co/s/n7m5kieb32uoc

**Comment by grcevski** [comment]
Failed again today: https://gradle-enterprise.elastic.co/s/ocxvyut2dr5cg/console-log?task=:x-pack:plugin:security:internalClusterTest

**Comment by DaveCTurner** [comment]
Failed again in my PR build: https://gradle-enterprise.elastic.co/s/kpanh6km54s56/console-log?task=:x-pack:plugin:security:internalClusterTest

**Comment by n1v0lg** [comment]
Phew, okay here is what I found:

**TL;DR** the test fails during the authentication step because a refresh token that is already available on a primary shard is not yet available on a replica shard

High level, the test fails under the following conditions:

Assume there are two nodes and the `.security-token` index has one primary shard and one replica

1. Thread `t1` attempts to refresh a token, and succeeds
1. The token doc is written to the primary shard and becomes available for Get requests
1. Thread `t2` attempts to refresh, finds there is a superseding token, fetches the superseding token from the primary shard and succeeds
1. Thread `t2` uses the refreshed token to authenticate, the request is routed to the replica where the token is not yet available. This leads to an authentication failure (and therefore a 401 among authStatuses here: https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java#L487)

Adding a `Thread.sleep()` to `IndexShard::applyIndexOperationOnReplica` to artificially delay the replica write helps reproduce the issue reliably.

I'll look into possible fixes and check in with @elastic/es-security next.

**Comment by ywangd** [comment]
This is [a known issue](https://github.com/elastic/elasticsearch/issues/55816#issuecomment-620913282). Unfortunately the original issue (#55816) was closed incorrectly because it was reused for 3 different failure scenarios. But this particular failure was never fixed for master. It was on mute until incorrectly [unmuted](https://github.com/elastic/elasticsearch/pull/85010/files#diff-9c0b337484d0450ee6109453bc351535a3118f25d5c3335b9fbd4d425cfd67ab) recently (also because the associated issue was incorrectly closed).

We had quite lengthy discussion on potential fixes. I can share more details in a separate channel.

**Comment by cbuescher** [comment]
fyi another one today: https://gradle-enterprise.elastic.co/s/wajlsuo3khpbk

**Comment by hendrikmuhs** [comment]
https://gradle-enterprise.elastic.co/s/ypguwuyqynrm2

**Comment by albertzaharovits** [comment]
Looks like some newly minted tokens failed auth.
I think it might have to do with exhausting some missing-doc retry count, but not 100% which one (there are several). I'll be sure after I can look at the trace logs of the TokenService.
I have raised https://github.com/elastic/elasticsearch/pull/100919.

**Comment by albertzaharovits** [comment]
It's low because the test is not actually muted, and the error scenario is convoluted and also covered by Kibana tests.

**Comment by ldematte** [comment]
Happened again today, in 8.10:
https://gradle-enterprise.elastic.co/s/bw6njig2pcepy/tests/task/:x-pack:plugin:security:internalClusterTest/details/org.elasticsearch.xpack.security.authc.TokenAuthIntegTests/testRefreshingMultipleTimesWithinWindowSucceeds?expanded-stacktrace=WyIwIl0&top-execution=1

**Comment by jfreden** [comment]
This test failed again and I did some investigation: 
Build scan: https://gradle-enterprise.elastic.co/s/zcwkbtbmmjsuw/tests/task/:x-pack:plugin:security:internalClusterTest/details/org.elasticsearch.xpack.security.authc.TokenAuthIntegTests/testRefreshingMultipleTimesWithinWindowSucceeds?top-execution=1

I'm not sure if this is the same issue as described above, but the assertion failure is the same.

In the logs before the test fails it shows the same access and refresh token being returned for all calls: 
```
[2023-12-13T06:50:33,874][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-269]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:33,911][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-256]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:33,946][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-260]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:33,949][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-273]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:33,955][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-259]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:33,960][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-276]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:33,967][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-267]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:33,981][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-257]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:33,983][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-271]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:33,984][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-270]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:33,985][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-258]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:33,989][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-261]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:33,989][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-264]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:33,989][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-254]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:34,003][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-255]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:34,006][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-265]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:34,007][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-275]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:34,008][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-274]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:34,009][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-266]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:34,016][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-268]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:34,018][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-272]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:34,018][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-263]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
[2023-12-13T06:50:34,020][INFO ][o.e.x.s.a.TokenAuthIntegTests] [[Thread-262]] received access token [gcSHBBjS2Cranfu3hw5VXvmcPBvEq5ZBOTxJ6c0AAAA=] and refresh token [gcSHBBgpeWt9KwDHXVjO7nvy996nq5ZBOTxJ6c0AAAA=]
```
The assertion then fails with: 
```
Expected: <1>
         but: was <2>
```

Since all access tokens returned are the same, the count should be 1, this leads me to belive there might some concurrency issue, or there is something wrong with the `stream().distinct().count()` code. I've changed the distinct count method and added logging to see if we can get some more info.

**Comment by maxhniebergall** [comment]
another failure: https://gradle-enterprise.elastic.co/s/qudnyq3q52m3q/tests/task/:x-pack:plugin:security:internalClusterTest/details/org.elasticsearch.xpack.security.authc.TokenAuthIntegTests/testRefreshingMultipleTimesWithinWindowSucceeds?top-execution=1

**Comment by jfreden** [comment]
The latest failure is for 8.11 and the fix/logging wasn't back ported unfortunately. Back porting the PR with https://github.com/elastic/elasticsearch/pull/104389 and https://github.com/elastic/elasticsearch/pull/104388.

**Comment by DiannaHohensee** [comment]
It looks like the changes from https://github.com/elastic/elasticsearch/pull/100919, adding debug logging, got [hit here](https://github.com/elastic/elasticsearch/blob/17bc83f481ddcfc032f6743a7a64c64000d55740/x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java#L636) in a new failure. Perhaps there's new information in the failure, @albertzaharovits.

[Here's the build scan failure](https://gradle-enterprise.elastic.co/s/qpeeivz357wfo/tests/task/:x-pack:plugin:security:internalClusterTest/details/org.elasticsearch.xpack.security.authc.TokenAuthIntegTests/testRefreshingMultipleTimesWithinWindowSucceeds?top-execution=1)
```
Jan 31, 2024 8:56:52 AM com.carrotsearch.randomizedtesting.RandomizedRunner$QueueUncaughtExceptionsHandler uncaughtException	
WARNING: Uncaught exception in thread: Thread[#709,Thread-30,5,TGRP-TokenAuthIntegTests]	
java.lang.AssertionError: unexpected	
	at __randomizedtesting.SeedInfo.seed([67D39B1CEA034557]:0)	
	at org.elasticsearch.test.ESTestCase.fail(ESTestCase.java:2127)	
	at org.elasticsearch.test.ESTestCase.fail(ESTestCase.java:2131)	
	at org.elasticsearch.xpack.security.authc.TokenAuthIntegTests.lambda$testRefreshingMultipleTimesWithinWindowSucceeds$15(TokenAuthIntegTests.java:636)	
	at java.base/java.lang.Thread.run(Thread.java:1583)	
Caused by: org.elasticsearch.client.ResponseException: method [GET], host [http://127.0.0.1:25367/], URI [/_security/_authenticate], status line [HTTP/1.1 401 Unauthorized]	
{"error":{"root_cause":[{"type":"security_exception","reason":"unable to authenticate with provided credentials and anonymous access is not allowed for this request","additional_unsuccessful_credentials":"oauth2 token: invalid token","header":{"WWW-Authenticate":["Basic realm=\"security\" charset=\"UTF-8\"","Bearer realm=\"security\"","ApiKey"]}}],"type":"security_exception","reason":"unable to authenticate with provided credentials and anonymous access is not allowed for this request","additional_unsuccessful_credentials":"oauth2 token: invalid token","header":{"WWW-Authenticate":["Basic realm=\"security\" charset=\"UTF-8\"","Bearer realm=\"security\"","ApiKey"]}},"status":401}	
	at org.elasticsearch.client.RestClient.convertResponse(RestClient.java:351)	
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:317)	
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:292)	
	at org.elasticsearch.test.TestSecurityClient.execute(TestSecurityClient.java:492)	
	at org.elasticsearch.test.TestSecurityClient.authenticate(TestSecurityClient.java:79)	
	at org.elasticsearch.xpack.security.authc.TokenAuthIntegTests.lambda$testRefreshingMultipleTimesWithinWindowSucceeds$15(TokenAuthIntegTests.java:634)	
	... 1 more	
Jan 31, 2024 8:56:52 AM com.carrotsearch.randomizedtesting.RandomizedRunner$QueueUncaughtExceptionsHandler uncaughtException	
WARNING: Uncaught exception in thread: Thread[#692,Thread-13,5,TGRP-TokenAuthIntegTests]	
java.lang.AssertionError: unexpected	
	at __randomizedtesting.SeedInfo.seed([67D39B1CEA034557]:0)	
	at org.elasticsearch.test.ESTestCase.fail(ESTestCase.java:2127)	
	at org.elasticsearch.test.ESTestCase.fail(ESTestCase.java:2131)	
	at org.elasticsearch.xpack.security.authc.TokenAuthIntegTests.lambda$testRefreshingMultipleTimesWithinWindowSucceeds$15(TokenAuthIntegTests.java:636)	
	at java.base/java.lang.Thread.run(Thread.java:1583)	
Caused by: org.elasticsearch.client.ResponseException: method [GET], host [http://127.0.0.1:25368/], URI [/_security/_authenticate], status line [HTTP/1.1 401 Unauthorized]	
{"error":{"root_cause":[{"type":"security_exception","reason":"unable to authenticate with provided credentials and anonymous access is not allowed for this request","additional_unsuccessful_credentials":"oauth2 token: invalid token","header":{"WWW-Authenticate":["Basic realm=\"security\" charset=\"UTF-8\"","Bearer realm=\"security\"","ApiKey"]}}],"type":"security_exception","reason":"unable to authenticate with provided credentials and anonymous access is not allowed for this request","additional_unsuccessful_credentials":"oauth2 token: invalid token","header":{"WWW-Authenticate":["Basic realm=\"security\" charset=\"UTF-8\"","Bearer realm=\"security\"","ApiKey"]}},"status":401}	
	at org.elasticsearch.client.RestClient.convertResponse(RestClient.java:351)	
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:317)	
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:292)	
	at org.elasticsearch.test.TestSecurityClient.execute(TestSecurityClient.java:492)	
	at org.elasticsearch.test.TestSecurityClient.authenticate(TestSecurityClient.java:79)	
	at org.elasticsearch.xpack.security.authc.TokenAuthIntegTests.lambda$testRefreshingMultipleTimesWithinWindowSucceeds$15(TokenAuthIntegTests.java:634)	
	... 1 more	
Jan 31, 2024 8:56:52 AM com.carrotsearch.randomizedtesting.RandomizedRunner$QueueUncaughtExceptionsHandler uncaughtException	
WARNING: Uncaught exception in thread: Thread[#695,Thread-16,5,TGRP-TokenAuthIntegTests]	
java.lang.AssertionError: unexpected	
	at __randomizedtesting.SeedInfo.seed([67D39B1CEA034557]:0)	
	at org.elasticsearch.test.ESTestCase.fail(ESTestCase.java:2127)	
	at org.elasticsearch.test.ESTestCase.fail(ESTestCase.java:2131)	
	at org.elasticsearch.xpack.security.authc.TokenAuthIntegTests.lambda$testRefreshingMultipleTimesWithinWindowSucceeds$15(TokenAuthIntegTests.java:636)	
	at java.base/java.lang.Thread.run(Thread.java:1583)	
Caused by: org.elasticsearch.client.ResponseException: method [GET], host [http://127.0.0.1:25366/], URI [/_security/_authenticate], status line [HTTP/1.1 401 Unauthorized]	
{"error":{"root_cause":[{"type":"security_exception","reason":"unable to authenticate with provided credentials and anonymous access is not allowed for this request","additional_unsuccessful_credentials":"oauth2 token: invalid token","header":{"WWW-Authenticate":["Basic realm=\"security\" charset=\"UTF-8\"","Bearer realm=\"security\"","ApiKey"]}}],"type":"security_exception","reason":"unable to authenticate with provided credentials and anonymous access is not allowed for this request","additional_unsuccessful_credentials":"oauth2 token: invalid token","header":{"WWW-Authenticate":["Basic realm=\"security\" charset=\"UTF-8\"","Bearer realm=\"security\"","ApiKey"]}},"status":401}	
	at org.elasticsearch.client.RestClient.convertResponse(RestClient.java:351)	
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:317)	
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:292)	
	at org.elasticsearch.test.TestSecurityClient.execute(TestSecurityClient.java:492)	
	at org.elasticsearch.test.TestSecurityClient.authenticate(TestSecurityClient.java:79)	
	at org.elasticsearch.xpack.security.authc.TokenAuthIntegTests.lambda$testRefreshingMultipleTimesWithinWindowSucceeds$15(TokenAuthIntegTests.java:634)	
	... 1 more
```

**Comment by kingherc** [comment]
One more failure at [elasticsearch / periodic / platform-support / 8.12 / opensuse-leap-15 / platform-support-unix](https://gradle-enterprise.elastic.co/s/y3yr6qm3adyco) similar to what @DiannaHohensee mentioned, with:

```
REPRODUCE WITH: ./gradlew ':x-pack:plugin:security:internalClusterTest' --tests "org.elasticsearch.xpack.security.authc.TokenAuthIntegTests.testRefreshingMultipleTimesWithinWindowSucceeds" -Dtests.seed=3EEA88B6A153A3B3 -Dtests.locale=nb-NO -Dtests.timezone=Africa/Dar_es_Salaam -Druntime.java=21

org.elasticsearch.xpack.security.authc.TokenAuthIntegTests > testRefreshingMultipleTimesWithinWindowSucceeds FAILED
    java.lang.AssertionError: unexpected
        at __randomizedtesting.SeedInfo.seed([3EEA88B6A153A3B3]:0)
        at org.elasticsearch.test.ESTestCase.fail(ESTestCase.java:2141)
        at org.elasticsearch.test.ESTestCase.fail(ESTestCase.java:2145)
        at org.elasticsearch.xpack.security.authc.TokenAuthIntegTests.lambda$testRefreshingMultipleTimesWithinWindowSucceeds$15(TokenAuthIntegTests.java:636)

        Caused by:
        org.elasticsearch.client.ResponseException: method [GET], host [http://127.0.0.1:24254], URI [/_security/_authenticate], status line [HTTP/1.1 401 Unauthorized]
        {"error":{"root_cause":[{"type":"security_exception","reason":"unable to authenticate with provided credentials and anonymous access is not allowed for this request","additional_unsuccessful_credentials":"oauth2 token: invalid token","header":{"WWW-Authenticate":["Basic realm=\"security\" charset=\"UTF-8\"","Bearer realm=\"security\"","ApiKey"]}}],"type":"security_exception","reason":"unable to authenticate with provided credentials and anonymous access is not allowed for this request","additional_unsuccessful_credentials":"oauth2 token: invalid token","header":{"WWW-Authenticate":["Basic realm=\"security\" charset=\"UTF-8\"","Bearer realm=\"security\"","ApiKey"]}},"status":401}
            at app//org.elasticsearch.client.RestClient.convertResponse(RestClient.java:351)
            at app//org.elasticsearch.client.RestClient.performRequest(RestClient.java:317)
            at app//org.elasticsearch.client.RestClient.performRequest(RestClient.java:292)
            at app//org.elasticsearch.test.TestSecurityClient.execute(TestSecurityClient.java:492)
            at app//org.elasticsearch.test.TestSecurityClient.authenticate(TestSecurityClient.java:79)
            at app//org.elasticsearch.xpack.security.authc.TokenAuthIntegTests.lambda$testRefreshingMultipleTimesWithinWindowSucceeds$15(TokenAuthIntegTests.java:634)
```

I see the test is muted in main, but not in 8.12. I will backport the mute in 8.12.

**Comment by mattc58** [comment]
@n1v0lg, look how old this ticket is. Is the test still valid, does it need to exist, etc.

**Comment by jfreden** [comment]
This test is still failing due to the scenario described [here](https://github.com/elastic/elasticsearch/issues/55816#issuecomment-713402239): 
>Say we have a 3-node cluster, the token index has its primary on node_A and replica on node_B and nothing on node_C. Thread 1 is refreshing the token and the new token document is processed by the primary, but has not made into the replica yet. Now Thread 2 comes to refresh the token as well, it hits node_C. Becuase node_C does not have a shard copy, it needs to route the request. Since the new token document has not been fully acknowledged yet, both primary and replica are considered equal candidate for the GET token document request. For the refresh request, it happens to be routed to the primary shard (node_A) and successfully get the document back. After the refresh, the new code attempts to authenticate with it. Although both the refresh and authenticate happen on the same node, the GET request can be routed to a different node. This time it gets routed to the replica shard (node_B). The replica shard does not have this document yet (because Thread 1 is still processing it) and so it fails.

And can be reproduced locally by doing what's suggested [here](https://github.com/elastic/elasticsearch/issues/85697#issuecomment-1120909065): 
>Adding a Thread.sleep() to IndexShard::applyIndexOperationOnReplica to artificially delay the replica write helps reproduce the issue reliably.

Having this test muted can cause issues, see [this slack thread](https://elastic.slack.com/archives/CA6N3HDF1/p1690388280716839) for example.

There are two suggested fixes mentioned [here](https://github.com/elastic/elasticsearch/issues/55816#issuecomment-713402239): 

1. Have TokenService#refreshToken to give back the token authentication object (Implemented in https://github.com/elastic/elasticsearch/pull/64178)
2. Retry the TokenService#authenticateToken (either in TransportRefreshTokenAction or within itself)

I will try to debug this further later this week as part of SDH duty.
